library/std/src/fs.rs

```diff
@@ -3333,26 +3333,39 @@ impl DirBuilder {
             return Ok(());
         }
 
-        match self.inner.mkdir(path) {
-            Ok(()) => return Ok(()),
-            Err(ref e) if e.kind() == io::ErrorKind::NotFound => {}
-            Err(_) if path.is_dir() => return Ok(()),
-            Err(e) => return Err(e),
-        }
-        match path.parent() {
-            Some(p) => self.create_dir_all(p)?,
-            None => {
-                return Err(io::const_error!(
-                    io::ErrorKind::Uncategorized,
-                    "failed to create whole tree",
-                ));
+        let ancestors = path.ancestors();
+        let mut uncreated_dir_ctr = 0;
+
+        for ancestor in ancestors {
+            if ancestor == Path::new("") {
+                break;
+            }
+
+            match self.inner.mkdir(ancestor) {
+                Ok(()) => break,
+                Err(e) if e.kind() == io::ErrorKind::NotFound => uncreated_dir_ctr += 1,
+                // we check if the err is AlreadyExists for two reasons
+                //    - in case the path exists as a *file*
+                //    - and to avoid calls to .is_dir() in case of other errs
+                //      (i.e. PermissionDenied)
+                Err(e) if e.kind() == io::ErrorKind::AlreadyExists && ancestor.is_dir() => break,
+                Err(e) => return Err(e),
             }
         }
-        match self.inner.mkdir(path) {
-            Ok(()) => Ok(()),
-            Err(_) if path.is_dir() => Ok(()),
-            Err(e) => Err(e),
+
+        // collect only the uncreated directories w/o letting the vec resize
+        let mut uncreated_dirs = Vec::with_capacity(uncreated_dir_ctr);
+        uncreated_dirs.extend(ancestors.take(uncreated_dir_ctr));
+
+        for uncreated_dir in uncreated_dirs.iter().rev() {
+            if let Err(e) = self.inner.mkdir(uncreated_dir) {
```

Comment:
```suggestion
                if e.kind() != io::ErrorKind::AlreadyExists || !uncreated_dir.is_dir() {
```

Here, you check `.is_dir()` even if you get an `io::ErrorKind::PermissionDenied`. With my suggestion, you'll only do it when the error ir `io::ErrorKind::AlreadyExists`.

Reply:
Oh okay, I understand where you're pointing to now (when you mentioned "below", I thought you meant the last case in the match statement). Yea, that's a completely fine check. This error checking was mainly for concurrent directory creations, but it could be possible to see some directory permission being changed concurrently or a file was created in this place instead of a directory.